### PR TITLE
Slots: use execution function return as a state arg value

### PR DESCRIPTION
**Pull request by DmitryKuzmenko**
### What does this PR do?
Adds an ability to use an execution function return as a state argument value.

### What issues does this PR fix or reference?
Implements #42136

### New Behavior
Allows to use `__slot__` like this:
```slsf
do-something:
  file.copy:
    - name: __slot__:salt:test.echo(/tmp/file-c)
    - source: __slot__:salt:test.echo(text=/etc/hosts)
```

### Tests written?
Yes

### Commits signed with GPG?
Yes

### CODE REVIEW

File: doc/topics/slots/index.rst [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+Slots
+=====
+
+.. versionadded:: Oxygen
+
+.. note:: This functionality is under development and could be changed in the
+    future releases
+
+Many times it is useful to store the results of a command during the course of
+an execution. Salt Slots is designed to allow to store this information and use
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: "is" should be "are" here

File: doc/topics/slots/index.rst [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+it later during the highstate or other job execution.
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: Can we link this to the highstate docs? For example:

```
:ref:`highstate <running-highstate>`
```

File: doc/topics/slots/index.rst [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+Slots extends the state syntax and allows to do things right before the state
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: "extends" should be "extend"

Comment (terminalmage) [reply doc/topics/slots/index.rst]: Also, can you change "allows to" to "allows you to"?

File: doc/topics/slots/index.rst [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+function be executed. So you can make a decision in the last moment right before
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: "be" should be "is"

File: doc/topics/slots/index.rst [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+a state is executed.
+
+Execution functions
+-------------------
+
+.. note:: Using execution modules return data as a state values is a first step
+    of Slots development. Other functionality is under development.
+
+Slots allow to use minion execution module return as an argument value in
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: Can you please re-word this as:

```
Slots allow you to use the return from a remote-execution function as an argument value in states.
```

File: doc/topics/slots/index.rst [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+states.
+
+Slot syntax looks close to the simple python function call.
+
+.. code-block::
+
+    __slot__:salt:<module>.<function>(<args>, ..., <kwargs...>, ...)
+
+
+Simple example is here:
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: Can you reword this as `Here is a simple example:` ?

File: doc/topics/slots/index.rst [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+.. code-block:: yaml
+
+    copy-some-file:
+      file.copy:
+        - name: __slot__:salt:test.echo(text=/tmp/some_file)
+        - source: __slot__:salt:test.echo(/etc/hosts)
+
+This will execute the `test.echo` execution functions right before calling the
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: We can make `test.echo` a link to the function documentation using:

```
:py:func:`test.echo <salt.modules.test.echo>`
```

File: doc/topics/slots/index.rst [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+
+Simple example is here:
+
+.. code-block:: yaml
+
+    copy-some-file:
+      file.copy:
+        - name: __slot__:salt:test.echo(text=/tmp/some_file)
+        - source: __slot__:salt:test.echo(/etc/hosts)
```
Comment (terminalmage) [inline doc/topics/slots/index.rst]: Is it correct that we do not need to quote the strings passed to these slot functions? If so, we should probably explicitly document that.

File: salt/state.py [hunk 9]
```diff
@@ -2039,6 +2040,41 @@ def call(self, low, chunks=None, running=None, retries=1):
...
+        # __slot__:salt.cmd.run(foo, bar, baz=qux)
+        ctx = ((u'args', enumerate(cdata[u'args'])),
+               (u'kwargs', cdata[u'kwargs'].items()))
+        for atype, avalues in ctx:
+            for ind, arg in avalues:
+                if not isinstance(arg, six.string_types) or not arg.startswith(u'__slot__:'):
+                    # Not a slot, skip it
+                    continue
+                cdata[atype][ind] = self.__eval_slot(arg)
+
```
Comment (terminalmage) [inline salt/state.py]: Sorry to ask this of you, but we're going to go with importing `unicode_literals` after all. I'll have a PR open for this later today. Can you remove the leading `u` from the string literals here?

**Review by terminalmage — CHANGES_REQUESTED** [review]
The code is stellar as usual. Nothing is technically wrong here, but there are some grammar modifications, a question about the docs, and a requested change to the unicode literals in salt/state.py, and then this should be good to merge.

File: doc/topics/slots/index.rst (hunk 8 above)
Comment (DmitryKuzmenko) [reply doc/topics/slots/index.rst]: Yes, it's correct. This is why I've wrote "syntax looks close". I'll add a note.

File: salt/state.py (hunk 9 above)
Comment (DmitryKuzmenko) [reply salt/state.py]: Sure, no problem!

Comment (DmitryKuzmenko) [reply salt/state.py]: No problem!

**Comment by DmitryKuzmenko** [comment]
@terminalmage big thanks for your review! I've applied all your suggestions. Could you please re-read the documentation part of the code?

**Review by thatch45 — APPROVED** [review]

**Review by terminalmage — APPROVED** [review]
Apologies for the delay. This looks good! Just needs a rebase before we can merge.

**Comment by rallytime** [comment]
@DmitryKuzmenko Looks like this needs a rebase. There a merge conflict now, unfortunately.

**Comment by DmitryKuzmenko** [comment]
@rallytime one commit has an expected conflict because of all `u''` were removed. So no logic touched. Could be merged.

**Review by cachedout — APPROVED** [review]

**Comment by rallytime** [comment]
@DmitryKuzmenko It looks like there are some related test failures on this change: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/18125/

**Comment by DmitryKuzmenko** [comment]
Hm... I'll check. Thank you @rallytime .

**Comment by acaiafa** [comment]
looking forward to this merge and completion.

**Comment by rallytime** [comment]
re-run py

**Comment by rallytime** [comment]
The tests that are failing are:
- setUpClass (unit.transport.test_zeromq.AESReqTestCases)
- setUpClass (unit.transport.test_zeromq.ClearReqTestCases)

This is on the Ubuntu 14 run: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/18475/

**Comment by DmitryKuzmenko** [comment]
@rallytime it looks very strange. It shouldn't be related to the changes.
The test log says "Can not allocate memory"...

**Comment by DmitryKuzmenko** [comment]
I've just added a release note and rebased this to the head of the oxygen branch. I think this should fix the test failure.

**Comment by DmitryKuzmenko** [comment]
re-run py

**Comment by rallytime** [comment]
Looks like the latest test run only has failures that are on the branch. 🎉

File: salt/state.py [hunk 10]
```diff
@@ -1930,6 +1930,7 @@ def call(self, low, chunks=None, running=None, retries=1):
                         # run the state call in parallel, but only if not in a prereq
                         ret = self.call_parallel(cdata, low)
                     else:
+                        self.format_slots(cdata)
```
Comment (max-arnold) [inline salt/state.py]: Is there any reason why `format_slots` isn't called for parallel states inside of `_call_parallel_target`?
